### PR TITLE
refactor!: item.search for json rpc works

### PR DESCRIPTION
**Pull request by f3fora**
item.search function has been simplified, eliminating some string manipulation (that can be eventually done on the "client" side).

The condition ('citationKey', 'contains', value) has been removed, since it was a too strong constrain.

The item.search accepts now a vector of tuple with 1 to 4 elements, as expected.

Other behaviors, previously present, has been deactivated, and some custom actions have been created to easily use them.

The library argument has been removed, since it is a special case for conditions.

### CODE REVIEW

**Comment by f3fora** [comment]
> Is it possible to have it as an `OR` condition? If so I'd prefer to keep that.

I think that encapsulating `block` to achieve the OR raises error. And the quicksearch already use one.

**Comment by mgmeyers** [comment]
Just noticed some issues with this on my end as well. This is indeed treated as an `AND` operation.

```js
for (const [name, operator, value] of terms) {
  search.addCondition(name, operator, value)
  search.addCondition('citationKey', 'contains', value)
}
```

I believe this can be used to make it `OR`:
```js
search.addCondition('joinMode', 'any')
```

But when I add this in, it seems to match my entire library when running this query:
```json
{
    "jsonrpc": "2.0",
    "method": "item.search",
    "params": [[["dateModified", "isAfter", "2023-12-21 22:40:13"]], 1]
}
```

**Comment by f3fora** [comment]
Is there anything needed from my end to facilitate the merge process of this PR?

**Comment by retorquere** [comment]
I had hoped we could address the and/or issue. I agree that default `and` is not useful behavior.

File: content/json-rpc.ts [hunk 1]
```diff
@@ -101,40 +101,86 @@ class NSUser {
...
+   * - search([['joinMode', 'any'], ['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero']]): search for entries with Creator 'Johnny' OR Title 'Zotero'
+   * - search([['blockStart'], ['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero'], ['blockEnd'], ['creator', 'contains', 'Smith']]): search for entries with (Creator 'Johnny' OR Title 'Zotero') AND (Creator 'Smith')
+   *
+   * @param terms  Single string as typed into the search box in Zotero (search for Title Creator Year)
+   *               Array of tuples similar as typed into the advanced search box in Zotero
+   *               (https://github.com/zotero/zotero/blob/9971f15e617f19f1bc72f8b24bb00b72d2a4736f/chrome/content/zotero/xpcom/data/searchConditions.js#L72-L610)
    */
-  public async search(terms: string | [string, string, string][], library?: string | number) {
+  public async search(terms: string
+  | ([string] | [string, string] | [string, string, string | number] | [string, string, string | number, boolean])[]) {
```
Comment (mgmeyers) [inline content/json-rpc.ts:124]: Should we still support passing the library ID as the second param? I think it's still needed for basic string searches.

File: content/json-rpc.ts [hunk 2]
```diff
@@ -101,40 +101,86 @@ class NSUser {
...
-        throw new Error(`library ${JSON.stringify(library)} not found`)
+    }
+    else {
+      blk: for (const term of terms) {
+        // Custom Actions
+        if ((term.length === 1)) {
+          switch (term[0]) {
+            case 'ignore_feeds': {
+              for (const feed of Zotero.Feeds.getAll()) {
+                search.addCondition('libraryID', 'isNot', feed.libraryID)
```
Comment (mgmeyers) [inline content/json-rpc.ts:168]: This should probably add `true` as the 4th parameter so these don't become an `OR` query if `joinMode` or `blockStart` are used.

File: content/json-rpc.ts [hunk 3]
```diff
@@ -101,40 +101,86 @@ class NSUser {
...
+   * - search('') or search([]): return every entries
+   * - search('Zotero'): quick search for 'Zotero'
+   * - search([['title', 'contains', 'Zotero']]): search for 'Zotero' in the Title
+   * - search([['library', 'is', 'My Library']]): search for entries in 'My Library'
+   *   (this function try to resolve the string 'My Library' into is own libraryId number)
+   * - search([['ignore_feeds']]): custom action for ignoring the feeds
+   * - search([['ignore_feeds'], ['quicksearch-titleCreatorYear', 'contains', 'Zotero']]): quick search for 'Zotero' ignoring the Feeds
+   * - search([['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero']]): search for entries with Creator 'Johnny' AND Title 'Zotero'
+   * - search([['joinMode', 'any'], ['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero']]): search for entries with Creator 'Johnny' OR Title 'Zotero'
+   * - search([['blockStart'], ['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero'], ['blockEnd'], ['creator', 'contains', 'Smith']]): search for entries with (Creator 'Johnny' OR Title 'Zotero') AND (Creator 'Smith')
```
Comment (mgmeyers) [inline content/json-rpc.ts:117]: From my understanding of the code, `blockStart` doesn't work this way. I could be wrong, though. It seems that `blockStart` is specifically used for quick search queries and has some specially handling in code for this use case. I think to acheive `OR` + `AND` the user would want to use `joinMode` `any` and also pass the `required` param: `[['joinMode', 'any'], ['creator', 'contains', 'Johnny'], ['title', 'contains', 'Zotero'], ['creator', 'contains', 'Smith', true]]`

**Review by mgmeyers — COMMENTED** [review]
Looks like updates to `site/themes` and `submodules` made it into your last commit. Otherwise, I think it's almost there!

**Pull request merged**